File: src/main/java/pl/edu/zut/mad/schedulecalendar/ScheduleFragment.java
package pl.edu.zut.mad.schedulecalendar;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

public class ScheduleFragment extends Fragment {

    private static final String CALENDAR_TAG = "calendar_fragment";
    private static final String SCHEDULE_TAG = "schedule_fragment";
    private static final String ABOUT_US_TAG = "about_us_fragment";
    private CalendarFragment calendarFragment;
    private SchedulePagerFragment schedulePagerFragment;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_schedule, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initScheduleFragments(savedInstanceState);
    }

    private void initScheduleFragments(Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            calendarFragment = new CalendarFragment();
            schedulePagerFragment = new SchedulePagerFragment();
            startScheduleFragments();
        } else {
            initScheduleFragmentsFromStack();
        }
        registerCalendarForScheduleFragment();
    }

    private void startScheduleFragments() {
        replaceFragmentInViewContainer(calendarFragment, R.id.calendar_container, CALENDAR_TAG);
        replaceFragmentInViewContainer(schedulePagerFragment, R.id.schedule_container, SCHEDULE_TAG);
    }

    private void replaceFragmentInViewContainer(Fragment fragment, int containerId, String tag) {
        getFragmentManager().beginTransaction()
                .replace(containerId, fragment, tag)
                .commit();
    }

    private void initScheduleFragmentsFromStack() {
        calendarFragment = (CalendarFragment) getFragmentFromStackWithTag(CALENDAR_TAG);
        schedulePagerFragment = (SchedulePagerFragment) getFragmentFromStackWithTag(SCHEDULE_TAG);
    }

    private Fragment getFragmentFromStackWithTag(String tag) {
        return getFragmentManager().findFragmentByTag(tag);
    }

    private void registerCalendarForScheduleFragment() {
        if (schedulePagerFragment != null) {
            schedulePagerFragment.registerCalendar(calendarFragment);
        }
    }

}

File: src/main/java/pl/edu/zut/mad/schedulecalendar/ScheduleFragment.kt
package pl.edu.zut.mad.schedulecalendar

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.ognev.kotlin.agendacalendarview.builder.CalendarContentManager
import com.ognev.kotlin.agendacalendarview.models.CalendarEvent
import kotlinx.android.synthetic.main.fragment_schedule.*
import pl.edu.zut.mad.schedulecalendar.login.LoginActivity
import pl.edu.zut.mad.schedulecalendar.module.ScheduleModule
import pl.edu.zut.mad.schedulecalendar.util.NetworkUtils
import pl.edu.zut.mad.schedulecalendar.util.app
import java.util.*
import javax.inject.Inject


class ScheduleFragment : Fragment(), ScheduleMvp.View {

    companion object {
        private val NETWORK_UTILS = NetworkUtils()
        private const val REQUEST_CODE = 123
    }

    @Inject lateinit var user: User
    @Inject lateinit var presenter: SchedulePresenter
    private val calendarContentManager: CalendarContentManager by lazy {
        CalendarContentManager(CalendarController(), scheduleCalendarView, LessonsAdapter(activity))
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
            inflater.inflate(R.layout.fragment_schedule, container, false)

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    private fun init() {
        initInjections()
        initView()
    }

    private fun initInjections() = app.component
            .plus(ScheduleModule(this))
            .inject(this)

    private fun initView() =
            if (user.isSaved()) {
                presenter.loadLessons()
            } else {
                startLoginActivity()
            }

    private fun startLoginActivity() {
        val intent = Intent(activity, LoginActivity::class.java)
        startActivityForResult(intent, REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK && requestCode == REQUEST_CODE) {
            presenter.loadLessons()
        }
    }

    override fun onDateIntervalCalculated(minDate: Calendar, maxDate: Calendar) {
        calendarContentManager.locale = Locale.getDefault()
        calendarContentManager.setDateRange(minDate, maxDate)
    }

    override fun onLessonsEventLoad(lessonsEvents: MutableList<CalendarEvent>) {
        calendarContentManager.loadItemsFromStart(lessonsEvents)
    }

    fun logout() {
        user.remove()
        startLoginActivity()
    }

    fun refreshSchedule() {
        if (NETWORK_UTILS.isAvailable(activity)) {
            startLoginActivityWithSavedAlbumNumber()
        } else {
            Toast.makeText(activity, R.string.error_no_internet, Toast.LENGTH_SHORT).show()
        }
    }

    private fun startLoginActivityWithSavedAlbumNumber() {
        val intent = Intent(activity, LoginActivity::class.java)
        // TODO: send action to refresh
        startActivity(intent)
    }
}
